Add "--spill" option to the refresh command that spills the content of the revision into worktree and empties its patch

NonLogicalDev opened the pull request:
In case a user wants to completely change what goes into the patch, they can run `stg refresh --spill` and then manually add/refresh files/changes back into the patch.

I feel like `stgit` currently lacks any easy way to achieve this.

This functionality closely mirrors what `git-delouse` script does:
[GitToolkit/GitDelouse](https://github.com/nvie/git-toolbelt#git-delouse)

ctmarinas commented:
Wouldn't this be better as an option to "stg refresh", say "--spill" or "--empty"? We have the equivalent "stg delete --spill" which deletes a patch and spills its content onto the worktree and index (well, it leaves them unchanged).

NonLogicalDev commented:
@ctmarinas I kind of like the `refresh --spill`, not too picky about semantics as long as this functionality is available in some form. I will refactor this accordingly. Thank you for you comments!

jpgrayson commented:
+1 for `refresh --spill`. It would be a good complement to `pop --spill` and `delete --spill`.

Thanks for taking on this new feature!

NonLogicalDev commented:
@jpgrayson Thank you so much for maintaining this project, I only discovered this this a few weeks ago, and I already can not live without it! Such a major quality of life improvement over my old way of working in a state of constant rebase in a stack of changes.

If you need help with anything in regards to this project I would be happy to lend a hand when time permits.

NonLogicalDev commented:
I refactored the `split` command to become a `refresh --spill` option.

A few thoughts:

The semantics of a spill on refresh command are a little clunky

As a first pass I have changed the refresh command implementation to branch out into `spill` function rather than trying to reconcile it with the rest of the refresh implementation. The clunkyness here is that refresh technically can operate on arbitrary patches in the stack, whereas the `--spill` option (as currently implemented by me) only operates at the topmost patch.

Do you think we should allow the user to pick which patch they might want to spill? Say if they had patches (A), (B), (C) in the stack and they have all 3 applied, would we want to support spilling patch (B) or (A)? Or should we for the sake of simplicity start with only supporting the topmost patch.

lkraav commented:
> Or should we for the sake of simplicity start with only supporting the topmost patch.

This makes sense to start with.

Spilling mid-stack patches has significant merge conflict risks, although it's basically the same as adding more stuff to a mid-stack patch.. So it's sort of 50/60 call :thinking: :smile:

NonLogicalDev commented:
Let's go with only the topmost patch for now. We can expand this to support arbitrary patches later without breaking backwards compatibility.

In this case if someone can review the changes, I will be super grateful.

jpgrayson reviewed (commented):
Overall this looks pretty good, but I've made note of a couple of nits.

Also, could you update `completion/stgit.zsh` with this new option?

Thanks again for this PR.

NonLogicalDev commented on stgit/commands/refresh.py:
```diff
@@ -342,6 +397,11 @@ def func(parser, options, args):
+    # Spill option acts a bit differently from the rest of the command flow.
+    # So it takes control from here.
+    if options.spill:
+        return spill(parser, options, args)
```
Noted. I refactored and took a slightly different approach for readability by breaking up the `func` body into option validation and two functions `__refresh_spill` and `__refresh` for two separate execution flows.

NonLogicalDev commented on stgit/commands/refresh.py:
```diff
@@ -318,6 +332,47 @@ def info_msg():
+    # Set the tree of the patch to the parent.
+    cd = cd.set_tree(cd.parent.data.tree)
+
+    iw = None
```
Fixed

NonLogicalDev commented on stgit/commands/refresh.py:
```diff
@@ -318,6 +332,47 @@ def info_msg():
+
+    iw = None
+    annotation = "spilling " + patchname
+    log_msg = 'refresh\n\n' + annotation
```
Fixed

NonLogicalDev commented on stgit/commands/refresh.py:
```diff
@@ -342,6 +397,11 @@ def func(parser, options, args):
         options.submodules = config.getbool('stgit.refreshsubmodules')
 
+    # Spill option acts a bit differently from the rest of the command flow.
+    # So it takes control from here.
```
Fixed

NonLogicalDev commented:
@jpgrayson I have incorporated your suggestions. Give this another look-over when you get a chance.

jpgrayson commented:
The refactors look good. Thanks for the extra effort.

Merged!
